### DISCUSSION TITLE
is  'tab.rows[:]=newData; tab.update()' not enought to update ui.table?

### DISCUSSION THREAD

**Discussion by retsyo** [zauberzeug/nicegui#2046] [Q&A] [answered]
### Question

The expected behaviour is simple: whenever a user clicks `submit`, 
- the `username`, `id` and `time` is send to the table's data
- the data in table is sorted by `time`
- the table should be updated immediately in every web-browsers, so that every user reads the same new record

however, my code does not run in that way. 
- The table on other web-browser rarely updates automatically. But if the user refresh his owen web-browser, he may find that the table can be updated with the new data
- atcually, I opened different webbrowsers on my PC to test. I think there is no such delay in the web connection  

so what is the problem? Thanks

```python
import time

from typing import Optional

from fastapi import Request
from fastapi.responses import RedirectResponse
from starlette.middleware.base import BaseHTTPMiddleware

from nicegui import Client, app, ui

passwords = {f'u{i}': f'p{i}' for i in range(1, 20)}

columns = [
    {'name': 'name', 'label': 'NAME', 'field': 'name'},
    {'name': 'id', 'label': 'ID', 'field': 'id'},
    {'name': 'time', 'label': 'TIME', 'field': 'time'},
]

rows = []
for name, id in passwords.items():
    rows.append({'name': name, 'id':  id, 'time':''})

tab = None

unrestricted_page_routes = {'/login'}

class AuthMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next):
        if not app.storage.user.get('authenticated', False):
            if request.url.path in Client.page_routes.values() and request.url.path not in unrestricted_page_routes:
                app.storage.user['referrer_path'] = request.url.path  # remember where the user wanted to go
                return RedirectResponse('/login')
        return await call_next(request)

app.add_middleware(AuthMiddleware)

def fnUpdateTab(name, id, t):
    global rows
    global tab
    for idx, line in enumerate(rows):
        if line['name']==name and line['id']==id :
            rows[idx] = {'name':name, 'id':id, 'time':t}
            break

    sortedList = sorted(rows, key=lambda x:x['time'], reverse=True)
    #~ rows = sortedList
    tab.rows[:] = sortedList
    tab.update()

def fnSubmit(name, id, t):
    fnUpdateTab(name, id, t)
    app.storage.user.clear()
    tab.update()
    login()

@ui.page('/')
def main_page(client:Client) -> None:
    localTime = time.localtime(time.time())
    strSt = time.strftime("%H:%M:%S", localTime)

    lstAllAnswerWidgets = []
    NAME = app.storage.user["username"]
    ID = passwords[NAME]
    ui.button('submit',
                on_click=lambda e: fnSubmit(NAME, ID, strSt))

@ui.page('/login')
def login() -> Optional[RedirectResponse]:
    global tab

    def try_login() -> None:  # local function to avoid passing username and password as arguments
        if passwords.get(username.value) == id.value:
            app.storage.user.update({'username': username.value, 'authenticated': True})
            ui.open(app.storage.user.get('referrer_path', '/'))  # go back to where the user wanted to go

    if app.storage.user.get('authenticated', False):
        return RedirectResponse('/')

    with ui.row():
        with ui.card():
            username = ui.input('name').on('keydown.enter', try_login)
            id = ui.input('id').on('keydown.enter', try_login)
            ui.button('login', on_click=try_login)

        tab = ui.table(columns=columns, rows=rows, pagination=5)


ui.run(storage_secret='THIS_NEEDS_TO_BE_CHANGED')
```

**Comment by natankeddem** [comment]
You example is really verbose, you should try to boil it down more if you can...
Also, I can't get your example to work? I never get a page with a submit button because of the middleware and removing that brings up other errors. It could also be I don't understand the example and everything is working as intended?

All that being said, some thoughts. You have your `ui.table` inside a routed page. You can think of that page as creating a unique instance of every element inside for every browser you open to that page. Using `update()` on a `ui.table` instance will only update the instance in that specific browser.

**Comment by nghia-vo** [comment]
You may find the answer here: https://github.com/zauberzeug/nicegui/discussions/1753. Try "tab._props['rows'] = sortedList"

**Comment by retsyo** [comment, marked as answer]
Thanks everyone. 

Just like what natankeddem and one of my local friend pointed out, I created individual `ui.table` per everyone. Although I declared `global tab`, it refers to different table for different logged in user. There are 2 ways to fix:
# 1. First solution
declare refreshable table which is presented for all users 
```python
@ui.refreshable
def login_table():
    return ui.table(
        ...
    )
```    
then do
```python
login_table.refresh()
```
if we need to update, which actually erases old table and creates a new one.

# Second solution
record all table for all users
```python
lstTab = []

@ui.page('/login')
def login() -> Optional[RedirectResponse]:

    ...
    tab = ui.table(columns=columns, rows=rows, pagination=5)

    lstTab.append(tab)
```
then we can use iterate to update all of them:
```python
rows = some_new_data

for tab in lstTab:
    tab.rows[:] = rows
    tab.update()
```

P.S.
The code seems too be verbose, but most of it is actually stolen from [authentication demo](https://github.com/zauberzeug/nicegui/blob/main/examples/authentication/main.py). So in other words, most deals with authentication problem, to my limited web programming experience, which is necessary in such case and can't be distilled. Well, please correct me if I am wrong.
